### PR TITLE
Refactor ExecutionContext by Separating into Static and Dynamic structs

### PR DESCRIPTION
**Pull request by danilowhk**
### Description
This pull request refactors the execution context within the codebase, focusing on optimizations that target the inherent read-only nature of Cairo's memory. The key challenge addressed is that in Cairo, each update to the execution context requires the entire struct to be recreated, leading to inefficiencies.

#### Changes
1. **Separation into Static and Dynamic Contexts**: To overcome the update inefficiency, the execution context has been separated into `StaticExecutionContext` and `DynamicExecutionContext`. The intention is to place both contexts within a Box pointer, allowing updates to be made to the pointer rather than the entire struct. This approach avoids the need to recopy the whole struct with each update.
    - **Note**: Currently, the `BoxTrait` implementation is facing some issues, so the optimization using Box pointers is pending. This PR lays the groundwork by separating the contexts but does not yet implement the Box pointers.

2. **Introduction of Specific Traits and Structs**: New traits and structs such as `StaticExecutionContextTrait`, `DynamicExecutionContextTrait`, `StaticExecutionContext`, and `DynamicExecutionContext` have been created to facilitate the separation of contexts.

3. **Update of Method Implementations**: Various method implementations have been refined to align with the new structure, enhancing both efficiency and code clarity.

4. **Test Adaptations**: Tests have been updated to match the new structure, ensuring that existing functionalities remain intact.

#### Impact
The refactor represents a vital step towards optimizing the execution context handling in Cairo. While the full optimization awaits the resolution of issues with the `BoxTrait`, this change sets the stage for that enhancement and provides immediate benefits in code maintainability and structure.

#### Review Request
I request a detailed review to ensure that the separation into static and dynamic contexts aligns with the intended design and that the changes are correctly implemented. Feedback on the planned use of Box pointers for future optimization is also welcome.

## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [X] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [X] Yes
- [ ] No

### CODE REVIEW

File: src/context.cairo [hunk 1]
```diff
@@ -60,31 +60,150 @@ impl DefaultCallContextImpl of Default<CallContext> {
...
+    fn evm_address(self: @StaticExecutionContext) -> EthAddress {
+        *self.evm_address
+    }
+
+    #[inline(always)]
+    fn is_read_only(self: @StaticExecutionContext) -> bool {
+        *self.read_only
+    }
+    #[inline(always)]
+    fn bytecode(self: @StaticExecutionContext) -> Span<u8> {
```
Comment (enitrat) [inline src/context.cairo:113]: this can be removed imo - As it's a part of the CallContext.

File: src/context.cairo [hunk 2]
```diff
@@ -60,31 +60,150 @@ impl DefaultCallContextImpl of Default<CallContext> {
...
+            create_addresses: Default::default(),
+            revert_contract_state: Default::default(),
+            return_data,
+            reverted: false,
+            stopped: false
+        }
+    }
+
+    #[inline(always)]
+    fn destroy_contracts(self: @DynamicExecutionContext) -> Span<EthAddress> {
```
Comment (enitrat) [inline src/context.cairo:144]: Just a note here - but unlike Rust we can't have `&ref mut`, so we can't really have a setter for these fields. Although, this is fine for a getter

Comment (enitrat) [reply src/context.cairo:144]: On a second note I think these methods should be defined on `ExecutionContext` directly and not on `DynamicExecutionContext`, which is just an code-optimisation pattern.

Comment (enitrat) [reply src/context.cairo:144]: That way we keep the "API" abstracted - our `ExecutionContext` has a specification we conform to, but let's not have too much non-important code

File: src/evm.cairo [hunk 3]
```diff
@@ -21,7 +21,7 @@ fn execute(
         call_context, starknet_address, evm_address, gas_limit, gas_price, returned_data, false
     );
     // Compute the intrinsic gas cost for the current transaction and increase the gas used.
-    ctx.process_intrinsic_gas_cost();
+    // ctx.process_intrinsic_gas_cost();
```
Comment (enitrat) [inline src/evm.cairo:24]: can we remove it totally already? or add TODO

File: src/interpreter.cairo [hunk 4]
```diff
@@ -53,7 +54,10 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
     fn decode_and_execute(ref self: EVMInterpreter, ref context: ExecutionContext) {
         // Retrieve the current program counter.
         let pc = context.program_counter;
-        let bytecode = (@context).call_context.bytecode(); //Note: here, bytecode returns a Span
+        let bytecode = (@context)
```
Comment (enitrat) [inline src/interpreter.cairo:57]: here we should only be calling `context.call_context.bytecode()`, abstracting the static/dynamic implementations

File: src/interpreter.cairo [hunk 5]
```diff
@@ -43,6 +43,7 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
         // Decode and execute the current opcode.
         self.decode_and_execute(ref context);
         // Check if the execution is complete.
+        // TODO: Check why context.dynamic_context.is_stopped() movess the context variable
```
Comment (enitrat) [inline src/interpreter.cairo:46]: probably because you access "move" dynamic_context when accessing it directly here. perhaps (@context.dynamic_context) would not cause any problem, to verify

**Comment by enitrat** [comment]
@danilowhk 
Just FYI, this error we had
```shell
thread 'main' panicked at 'Internal error, cycle detected:

DatabaseKeyIndex { group_index: 5, query_index: 16, key_index: 1 }
DatabaseKeyIndex { group_index: 5, query_index: 15, key_index: 2 }
DatabaseKeyIndex { group_index: 5, query_index: 14, key_index: 2 }
```

appears when we do 
```rust
    let mut ctx = ExecutionContextTrait::new(
        call_context, starknet_address, evm_address, gas_limit, gas_price, returned_data, false
    );
 ```
 
 in `evm.cairo`

**Comment by enitrat** [comment]
After further investigation I isolated the bug which comes from a `Derive(Destruct)` instruction. I will report it to the compiler team. 

```rust
#[derive(Destruct)]
struct MinimalStruct {}

#[generate_trait]
impl MinimalStructImpl of MinimalStructTrait {
    fn new() -> MinimalStruct {
        MinimalStruct {}
    }
}

fn execute() {
    let mut ctx = MinimalStructTrait::new();
}


```

File: src/tests/test_execution_context.cairo [hunk 6]
```diff
@@ -69,36 +72,32 @@ fn test_execution_context_new() {
 
     // Then
     let call_context = setup_call_context();
-    assert(execution_context.call_context == call_context, 'wrong call_context');
+    assert(execution_context.static_context.call_context == call_context, 'wrong call_context');
```
[the remark this answers is not in the document]
Comment (Eikix) [reply src/tests/test_execution_context.cairo:75]: Discussed in Daily: For callcontext, it is the edge case `execution_context.call_context().calldata()`

[the remark this answers is not in the document]
Comment (Eikix) [reply src/tests/test_execution_context.cairo:75]: But for dynamic and static-> directly implement the getter: execution_context.pc()

File: src/context.cairo [hunk 7]
```diff
@@ -12,7 +12,7 @@ use option::OptionTrait;
 use starknet::get_caller_address;
 
 /// The call context.
-#[derive(Destruct)]
+#[derive(Destruct, Copy)]
```
Comment (enitrat) [inline src/context.cairo:15]:
```suggestion
#[derive(Drop, Copy)]
```

File: src/context.cairo [hunk 8]
```diff
@@ -60,31 +60,150 @@ impl DefaultCallContextImpl of Default<CallContext> {
 }
 
 
-/// The execution context.
-/// Stores all data relevant to the current execution context.
-#[derive(Destruct)]
-struct ExecutionContext {
+#[derive(Destruct, Copy)]
```
Comment (enitrat) [inline src/context.cairo:63]:
```suggestion
#[derive(Drop, Copy)]
```

File: src/context.cairo [hunk 9]
```diff
@@ -60,31 +60,150 @@ impl DefaultCallContextImpl of Default<CallContext> {
...
-    //sub_context: Nullable<ExecutionContext>,
+    read_only: bool,
+}
+
+impl DefaultStaticExecutionContext of Default<StaticExecutionContext> {
+    fn default() -> StaticExecutionContext {
+        StaticExecutionContext {
+            call_context: Default::default(),
+            starknet_address: 0.try_into().unwrap(),
+            evm_address: 0.try_into().unwrap(),
```
Comment (Eikix) [inline src/context.cairo:76]: I think we should implement Default trait for these types rather

Comment (Eikix) [reply src/context.cairo:76]: Or even implement a AddressZero trait method

File: src/interpreter.cairo (hunk 5 above)
Comment (danilowhk) [reply src/interpreter.cairo:46]: It worked with:

        if !((@context).dynamic_context.is_stopped()) {
            // Execute the next opcode.
            self.run(ref context);
        }
        
But in any case have changed to context.is_stopped

File: src/context.cairo (hunk 9 above)
Comment (danilowhk) [reply src/context.cairo:81]: Implemented default traits that will default to 0 addresses

File: src/context.cairo [hunk 10]
```diff
@@ -164,74 +216,127 @@ impl ExecutionContextImpl of ExecutionContextTrait {
...
+        *self.dynamic_context.stopped
+    }
+
+    #[inline(always)]
+    fn call_context(self: @ExecutionContext) -> CallContext {
+        *self.static_context.call_context
+    }
+
+    #[inline(always)]
+    fn destroy_contracts(self: @ExecutionContext) -> Span<EthAddress> {
```
Comment (enitrat) [inline src/context.cairo:255]: `destroy` is an active verb that implies and action. If we're getting a view, maybe`destroyed_contracts?`

```suggestion
    fn destroyed_contracts(self: @ExecutionContext) -> Span<EthAddress> {
```

File: src/context.cairo [hunk 11]
```diff
@@ -164,74 +216,127 @@ impl ExecutionContextImpl of ExecutionContextTrait {
...
+        self.dynamic_context.events.span()
+    }
+
+    #[inline(always)]
+    fn create_addresses(self: @ExecutionContext) -> Span<EthAddress> {
+        self.dynamic_context.create_addresses.span()
+    }
+
+    #[inline(always)]
+    fn revert_contract_state(self: @ExecutionContext) -> @Felt252Dict<felt252> {
```
Comment (enitrat) [inline src/context.cairo:270]: I think that's the only case where we will need to access it without getters - as a snapshot to a Dict is basically useless

Comment (enitrat) [reply src/context.cairo:270]: you can delete this one

File: src/interpreter.cairo [hunk 12]
```diff
@@ -54,7 +54,7 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
     fn decode_and_execute(ref self: EVMInterpreter, ref context: ExecutionContext) {
         // Retrieve the current program counter.
         let pc = context.program_counter;
-        let bytecode = (@context).call_context.bytecode(); //Note: here, bytecode returns a Span
+        let bytecode = (@context).call_context().bytecode(); //Note: here, bytecode returns a Span
```
Comment (enitrat) [inline src/interpreter.cairo:57]:
```suggestion
        let bytecode = context.call_context().bytecode(); //Note: here, bytecode returns a Span
```

no need to explicitly @ here, its done by the compiler

File: src/context.cairo (hunk 9 above)
Comment (enitrat) [reply src/context.cairo]: 0 for default sounds good for me tbh

File: src/context.cairo (hunk 10 above)
Comment (Eikix) [reply src/context.cairo:255]: or selfdestruct_contracts for even more explicit usecase

Comment (danilowhk) [reply src/context.cairo:255]: Went with destroyed_contracts as it felt cleaner

**Review by enitrat — APPROVED** [review]
LgTM
